File: .tmp_py4013/batch_12_dump/3309_maximum_possible_number_by_binary_concatenation.js
// LeetCode 3309 - Maximum Possible Number by Binary Concatenation
// https://leetcode.com/problems/maximum-possible-number-by-binary-concatenation/

function toBin(x) {
    if (x === 0) return '0';
    let s = '';
    while (x > 0) {
        s = String(x & 1) + s;
        x >>= 1;
    }
    return s;
}
function perm(i, idx, bs, ans) {
    if (i === 3) {
        const s = bs[idx[0]] + bs[idx[1]] + bs[idx[2]];
        let v = 0;
        for (const c of s) v = v * 2 + (c.charCodeAt(0) - 48);
        if (v > ans[0]) ans[0] = v;
        return;
    }
    for (let j = i; j < 3; j++) {
        let t = idx[i]; idx[i] = idx[j]; idx[j] = t;
        perm(i + 1, idx, bs, ans);
        t = idx[i]; idx[i] = idx[j]; idx[j] = t;
    }
}
var maxGoodNumber = function(nums) {
    const bs = [toBin(nums[0]), toBin(nums[1]), toBin(nums[2])];
    const idx = [0, 1, 2];
    const ans = [0];
    perm(0, idx, bs, ans);
    return ans[0];
};

File: .tmp_rs4013/rust_check16/3309_maximum_possible_number_by_binary_concatenation.rs
struct Solution;
// LeetCode 3309 - Maximum Possible Number by Binary Concatenation
// https://leetcode.com/problems/maximum-possible-number-by-binary-concatenation/

impl Solution {
    fn to_bin(mut x: i32) -> String {
        if x == 0 {
            return "0".to_string();
        }
        let mut s = String::new();
        while x > 0 {
            s.insert(0, if x & 1 == 1 { '1' } else { '0' });
            x >>= 1;
        }
        s
    }

    pub fn max_good_number(nums: Vec<i32>) -> i32 {
        let bs = [Self::to_bin(nums[0]), Self::to_bin(nums[1]), Self::to_bin(nums[2])];
        let mut idx = [0, 1, 2];
        let mut ans = 0;
        fn perm(i: usize, idx: &mut [usize; 3], bs: &[String; 3], ans: &mut i32) {
            if i == 3 {
                let s = format!("{}{}{}", bs[idx[0]], bs[idx[1]], bs[idx[2]]);
                let mut v = 0;
                for c in s.bytes() {
                    v = v * 2 + (c - b'0') as i32;
                }
                if v > *ans {
                    *ans = v;
                }
                return;
            }
            for j in i..3 {
                idx.swap(i, j);
                perm(i + 1, idx, bs, ans);
                idx.swap(i, j);
            }
        }
        perm(0, &mut idx, &bs, &mut ans);
        ans
    }
}

fn main() {}
